specs/eid-hermes-commands-format.md

```diff
@@ -0,0 +1,227 @@
+| 31:08 | Command Specific                         |
+
+**Table 1: Command Request format**
+
+| Bytes | Description        |
+|-------|--------------------|
+| 01:00 | Command Identifier |
+| 02    | Status             |
+| 07:03 | Reserved           |
+| 15:08 | Command Specific   |
+
+**Table 2: Command Response format**
+
+The following `Status` of Command Responses are defined:
+
+| Status | Description          |
+|--------|----------------------|
+| 0x00   | Success              |
+| 0x01   | Not enough space     |
+| 0x02   | Invalid Program Slot |
+| 0x03   | Invalid Data Slot    |
+| 0x04   | Invalid Address      |
+| 0x05   | eBPF error           |
+
+**Table 3: Status**
+
+## List of Commands
+
+The following commands are defined for Hermes devices:
+
+| Opcode | Command        |
+|--------|----------------|
+| 0x00   | Request Slot   |
+| 0x01   | Release Slot   |
+| 0x10   | Write to Slot  |
+| 0x11   | Read from Slot |
+| 0x80   | Run Program    |
+
+**Table 4: List of commands**
+
+### Request Slots (Opcode 0x00)
+
+This command can be used to reserve a program or data slot on the device.
+
+| Bytes | Description                           |
+|-------|---------------------------------------|
+| 08    | Slot type (0x00: program, 0x01: data) |
+
+**Table 5: Request Slots Command Request**
+
+| Bytes | Description                                          |
+|-------|------------------------------------------------------|
+| 08    | ID of allocated slot. Only valid when Status is 0x00 |
+
+**Table 6: Request Slots Command Response**
+
+This command may return the following status:
+
+| Status | Description                            |
+|--------|----------------------------------------|
+| 0x00   | Success                                |
+| 0x01   | All slots of requested type are in use |
+
+**Table 7: Request Slots status codes**
+
+### Release Slots (Opcode 0x01)
+
+This command can be used to release a program or data slot on the device.
+
+It is invalid to release a slot that is not allocated.
+
+| Bytes | Description                           |
+|-------|---------------------------------------|
+| 08    | Slot type (0x00: program, 0x01: data) |
+| 09    | Slot ID                               |
+
+**Table 8: Release Slots Command Request**
+
+| Bytes | Description                |
+|-------|----------------------------|
+
+(no command-specific bytes used)
+
+**Table 9: Release Slots Command Response**
+
+This command may return the following status:
+
+| Status | Description                                                |
+|--------|------------------------------------------------------------|
+| 0x00   | Success                                                    |
+| 0x02   | Requested program slot does not exist or is not allocated  |
+| 0x03   | Requested data slot does not exist or is not allocated     |
+
+**Table 10: Release Slots status codes**
+
+### Write to Slot (Opcode 0x10)
```

Comment:
Don't we use the XDMA to write to a data slot? I do not think we want to re-implement that functionality via the hermes interface?

Reply:
We will, but the idea would be that the hermes driver would call the relevant xdma functions. There are a couple of advantages to this:
- We can do some validation on the input (such as valid length, the slot is allocated to this user, etc.)
- The user doesn't have to worry about computing the BAR4 offsets for the DMA

Reply:
Oh but @iomartin I though we were defining the host<->device interface. There will be a different definition for the userspace<->driver interface will tthere not?

Reply:
@iomartin as discussed lets make it clear when we are referring to the host<->device spec and when we are referring to the userspace<->driver spec. And yes we can resuse the same formats for the non-xdma commands....